python/ray/log_monitor.py

```diff
@@ -19,91 +19,191 @@
+            files.
+        can_open_more_files (bool): True if we can still open more files and
+            false otherwise.
     """
 
-    def __init__(self,
-                 redis_ip_address,
-                 redis_port,
-                 node_ip_address,
-                 redis_password=None):
+    def __init__(self, logs_dir, redis_address, redis_password=None):
         """Initialize the log monitor object."""
-        self.node_ip_address = node_ip_address
-        self.redis_client = redis.StrictRedis(
-            host=redis_ip_address, port=redis_port, password=redis_password)
-        self.log_files = {}
-        self.log_file_handles = {}
-        self.files_to_ignore = set()
+        self.host = os.uname()[1]
+        self.logs_dir = logs_dir
+        self.redis_client = ray.services.create_redis_client(
+            redis_address, password=redis_password)
+        self.log_filenames = set()
+        self.open_file_infos = []
+        self.closed_file_infos = []
+        self.can_open_more_files = True
+
+    def close_all_files(self):
+        """Close all open files (so that we can open more)."""
+        while len(self.open_file_infos) > 0:
+            file_info = self.open_file_infos.pop(0)
+            file_info.file_handle.close()
+            file_info.file_handle = None
+            self.closed_file_infos.append(file_info)
+        self.can_open_more_files = True
 
     def update_log_filenames(self):
-        """Get the most up-to-date list of log files to monitor from Redis."""
-        num_current_log_files = len(self.log_files)
-        new_log_filenames = self.redis_client.lrange(
-            "LOG_FILENAMES:{}".format(self.node_ip_address),
-            num_current_log_files, -1)
-        for log_filename in new_log_filenames:
-            logger.info("Beginning to track file {}".format(log_filename))
-            assert log_filename not in self.log_files
-            self.log_files[log_filename] = []
-
-    def check_log_files_and_push_updates(self):
-        """Get any changes to the log files and push updates to Redis."""
-        for log_filename in self.log_files:
-            if log_filename in self.log_file_handles:
-                # Get any updates to the file.
-                new_lines = []
-                while True:
-                    current_position = (
-                        self.log_file_handles[log_filename].tell())
-                    next_line = self.log_file_handles[log_filename].readline()
-                    if next_line != "":
-                        new_lines.append(next_line)
-                    else:
-                        self.log_file_handles[log_filename].seek(
-                            current_position)
-                        break
+        """Update the list of log files to monitor."""
+        try:
+            log_filenames = os.listdir(self.logs_dir)
+        except (IOError, OSError) as e:
+            if e.errno == errno.EMFILE:
```

Comment:
might want to use a limit here instead of going all the way up to the max number of files, bad things might happen if we hit that limit